docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+
+Many devils live in the details. The rest of this chapter will discuss the
+numerous edge cases and implementation concerns that Lucet's asynchronous
+signal implementation must consider, and arguments for its correctness in the
+face of these.
+
+First, a flow chart of the various states and their transitions:
+![state flow chart](states.png)
+
+This graph describes the various states that reflect values of
+`execution_domain`, and includes some timeout transitions reflecting a timeout
+occurring entirely in a specific domain, well away from racing with other
+machinery in wasm guest. The rest of this section discusses the correcness of
+timeouts at boundaries where these state transitions occur.
+
+For reference later, the possible state transitions are:
+* `A -> B`
+* `A -> D`
+* `B -> C`
+* `B -> E` (normal guest exit)
+* `B -> E` (from a guest fault/timeout)
+* `C -> B`
+* `C -> E` (hostcall terminates instance)
+* `C -> E` (hostcall observes timeout)
+  - not an internal state but we will also discuss timeouts during a hostcall fault
+* `D -> E`
+
+### `A -> B` timeout
+
+If a timeout occurs between instance initialization (A) and the start of guest
+execution (B), there are two circumstances to consider: does the timeout occur
+before or after the Lucet instance has switched to `Domain::Guest`?
+
+#### Before switching to `Domain::Guest`
+
+The `KillSwitch` that fires acquires `terminable` and then locks
+`execution_domain` to determine timeout mechanism. This is before the instance
+has locked it in `enter_guest_region`, so it will acquire the lock, with a
+state of `Domain::Pending`. Seeing a pending instance, the `KillSwitch` will
+update it to `Domain::Cancelled` and release the lock, at which point the
+instance will acquire the lock, observe the instance is `Cancelled`, and return
+to the host without executing any guest code.
+
+#### After switching to `Domain::Guest`
+
+The `KillSwitch` that fires acquires `terminable` and then attempts to acquire
+a lock on `execution_domain` to determine timeout mechanism. Because the
+instance has already locked `execution_domain` to update it to `Domain::Guest`,
+this blocks until the instance releases the lock (and guest code is running).
+At this point, the instance is running guest code and it is safe for the
+`KillSwitch` to operate as if it were terminating any other guest code - with
+the same machinery as an instance in state `B` (a `SIGALRM`).
+
+### `A -> D` timeout
+
+This is a timeout while another `KillSwitch` has already fired, timing out a
+guest before exeuction. Because another `KillSwitch` must have fired for there
+to be a race, one of the two will acquire `terminable` and actually update
+`execution_domain`, while the other simply exits.
+
+### `B -> C` timeout
+
+If a timeout occurs during a transition from guest (B) to hostcall (C) code,
+there are two circumstances also contingent on if the instance has or has not
```

Comment:
Maybe, "...on whether the instance has switched to `Domain::Hostcall`.

Reply:
`whether` is the word I was looking for! Thank you.